SALTO-6040: Add service placeholder config doc (#6025)
salto-io/salto: packages/serviceplaceholder-adapter/config_doc.md

# serviceplaceholder configuration

## Configuration example

```hcl
serviceplaceholder {
  client = {
    retry = {
      maxAttempts = 5
      retryDelay = 5000
    }
    rateLimit = {
      total = -1
      get = 20
    }
  }
  fetch = {
    include = [
      {
        type = ".*"
      },
    ]
    exclude = [
      {
        type = "Group" // TODO replace with a service type
      }
    ]

>   }

Reviewer: nit: maybe also add an example of defining custom elem ids (can use something like `myType` so that we don't need to change these per adapter)

#### Client timeout configuration options

| Name               | Default when undefined | Description                                                                                                            |
| ------------------ | ---------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| maxDuration        | `0` (unlimited)        | Set a timeout (in milliseconds) on requests                                                                            |
| retryOnTimeout     | true                   | Whether to retry requests that returned a timeout response                                                             |
| lastRetryNoTimeout | true                   | Whether to disable the timeout duration on the last retry (if we assume the service will eventually return a response) |

#### Client retry options

| Name        | Default when undefined | Description                                      |
| ----------- | ---------------------- | ------------------------------------------------ |

> | maxAttempts | `5` // TODO adjust     | The number of attempts to make for each request  |

Reviewer: nit: maybe add a single TODO in the beginning saying all numbers and values should be adjusted?

| [exclude](#fetch-entry-options)     | []                     | List of entries to determine what instances to exclude in the fetch |
| [elemID](#element-id-customization) | {}                     | Allows customizing element IDs for specific types                   |

## Fetch entry options

| Name                              | Default when undefined | Description                                                     |
| --------------------------------- | ---------------------- | --------------------------------------------------------------- |
| type                              | ""                     | A regex of the Salto type name to include in the entry          |
| [criteria](#fetch-entry-criteria) |                        | A List of criteria to filter specific instance of certain types |

## Fetch entry criteria

| Name   | Default when undefined | Description                                                                  |
| ------ | ---------------------- | ---------------------------------------------------------------------------- |

> | type   | .\*                    | A regex used to filter instances by matching the regex to their type name    |

Reviewer: nit: 
```suggestion
| type   | .\*                    | A regular expression used to filter instances by matching to their type name    |
```

> ## Element ID parts definitions

Reviewer: I think this should come after the Elment ID customization section, no?

| Name        | Default when undefined | Description                                                                         |
| ----------- | ---------------------- | ----------------------------------------------------------------------------------- |

> | fieldName   | system's default       | The instance's field name value will be used to create the element ID               |

Reviewer: ```suggestion
| fieldName   | system's default       | The instance's field name whose value will be used to create the element ID               |
```

| isReference | false                  | Set to true if the fieldName is also a Salto reference                              |
| mapping     | undefined              | Allows applying a specific function to fieldName. Options: 'uppercase', 'lowercase' |

## Element ID customization

| Name                                   | Default when undefined | Description                                                                  |
| -------------------------------------- | ---------------------- | ---------------------------------------------------------------------------- |
| [parts](#element-id-parts-definitions) | system's default       | All parts definitions will be concatenated to create the element ID          |

> | extendsParent                          | false                  | Whether to concat the parent element ID with the current element ID          |

Reviewer: ```suggestion
| extendsParent                          | false                  | Whether to concatenate the parent element ID with the current element ID          |
```

> | delimiter                              | "\_"                   | The delimiter to use to concat Elem ID "parts"                               |

Reviewer: ```suggestion
| delimiter                              | "\_"                   | The delimiter to use to concatenate Elem ID "parts"                               |
```

> | extendSystemPartsDefinition            | false                  | Set to true to combine system's parts definitions with user part definitions |

Reviewer: ```suggestion
| extendSystemPartsDefinition            | false                  | Set to true to append the user-defined parts with the default system definition. When false, the user definitions override the system definitions |
```

Author: yap much clearer 😅
